implementation-guide/Implementation Guide.md

```diff
@@ -0,0 +1,632 @@
+```
+
+#### Depositing the gift
+
+![](./handoffs-3party2.webm)
+
+This is done by the gifter, in our example Alisha's node, Node A. Each node needs a per-session gifting table which stores a gift ID to some reference to an object or promise. The gift ID is just a positive integer (usually incrementing integer) which is unique to the session which represents the reference being shared. When a handoff is initiated it's to a specific object on a node, not just the Node, the gifting mechanism is providing a secure way to share a specific reference to the receiver. The way a node deposits a gift is on the bootstrap object which has a specific `deposit-gift` method.
+
+Alisha starts implementing this by creating a gift table for each session and an auto incrementing counter to track the gift IDs she's sending. Alisha also implements the `deposit-gift` method on her bootstrap object which takes two arguments a gift ID and a reference and she puts these things into her gift table she's just made. Alisha then starts by adding code to her `op:deliver` and `op:deliver-only` to look through all the arguments being sent and when it encounters a reference to an object initialize a handoff and replace that reference with a `desc:handoff-give`. When a handoff is initialized in Alisha's system, it gets the current next gift ID and sends a message to the deposit gift method on the bootstrap object, this looks like this:
+
+```
+;; Deposit the object Node C is exporting at position `1` at gift ID `5`
+<op:deliver-only <desc:export 0> ['deposit-gift 5 <desc:export 1>]
+```
+ 
+#### Creating the `desc:handoff-give` certificate
+
+![](./handoffs-3party3.webm)
+
+Alisha next needs to create the `desc:handoff-give` certificate which she can give to Ben on Node B which will enable him to retrieve the reference Alisha left as a gift on Node C. The certificate has the following structure:
+
+```
+<desc:handoff-give receiver-key       ; the public key of the receiver in the gifter <-> receiver session
+                   exporter-location  ; OCapN Locator for the exporter
+                   session            ; The session ID for the gifter <-> exporter session
+                   gifter-side        ; the public key of the gifter in the gifter <-> exporter session
+                   gift-id>           ; Positive Integer or zero.
+```
+
+This includes a lot of information, some of which comes from the Gifter's session with the Receiver and some of it comes from the Gifter's session with the Exporter. It's important to keep in mind when thinking about handoffs is that it does not rely on keeping secrets, all messages sent can be public while remaining secure and handoffs are capabilities which only allow the receiver to obtain a reference to the object that the gifter deposits. The burdon for checking the information within the certificates rests with the exporter since it is the Node which has the reference to provide.
```

Comment:
```suggestion
This includes a lot of information, some of which comes from the Gifter's session with the Receiver and some of it comes from the Gifter's session with the Exporter. It's important to keep in mind when thinking about handoffs is that it does not rely on keeping secrets, all messages sent can be public while remaining secure and handoffs are capabilities which only allow the receiver to obtain a reference to the object that the gifter deposits. The burden for checking the information within the certificates rests with the exporter since it is the Node which has the reference to provide.
```